src/node/internal/internal_tls_wrap.ts

```diff
@@ -0,0 +1,644 @@
+  this[kIsVerified] = false;
+  this[kPendingSession] = null;
+  this[kErrorEmitted] = false;
+
+  let wrap: Socket | null = null;
+  let handle: Socket['_handle'] | null = null;
+  let wrapHasActiveWriteFromPrevOwner = false;
+
+  if (socket) {
+    if (socket instanceof Socket && socket._handle) {
+      wrap = socket;
+      // TODO(soon): Check TLS connection type and call starttls() if needed.
+    } else {
+      // Cloudflare Workers does not support any other socket type.
+      throw new ERR_OPTION_NOT_IMPLEMENTED('options.socket');
+    }
+
+    handle = wrap._handle;
+    wrapHasActiveWriteFromPrevOwner = wrap.writableLength > 0;
+  }
+
+  // Just a documented property to make secure sockets
+  // distinguishable from regular ones.
+  this.encrypted = true;
+
+  Reflect.apply(Socket, this, [
+    {
+      handle: this._wrapHandle(wrap, handle, wrapHasActiveWriteFromPrevOwner),
+      allowHalfOpen: socket ? socket.allowHalfOpen : tlsOptions.allowHalfOpen,
+      pauseOnCreate: tlsOptions.pauseOnConnect,
+      manualStart: true,
+      highWaterMark: tlsOptions.highWaterMark,
+      onread: !socket ? tlsOptions.onread : null,
+      signal: tlsOptions.signal,
+    } as SocketOptions,
+  ]);
+
+  // Proxy for API compatibility
+  this.ssl = this._handle; // C++ TLSWrap object
+
+  this.on('error', this._tlsError.bind(this));
+}
+Object.setPrototypeOf(TLSSocket.prototype, Socket.prototype);
+Object.setPrototypeOf(TLSSocket, Socket);
+
+TLSSocket.prototype.disableRenegotiation = function disableRenegotiation(
+  this: TLSSocket
+): void {
+  // Do nothing.
+};
+
+TLSSocket.prototype._wrapHandle = function _wrapHandle(
+  this: TLSSocket,
+  wrap: null | Socket,
+  handle: Socket['_handle'] | null | undefined,
+  _wrapHasActiveWriteFromPrevOwner: boolean
+): unknown {
+  if (!handle) {
+    return null;
+  }
+  this[kRes] = handle;
+
+  // Guard against adding multiple listeners, as this method may be called
+  // repeatedly on the same socket by reinitializeHandle
+  if (this.listenerCount('close', onSocketCloseDestroySSL) === 0) {
+    this.once('close', onSocketCloseDestroySSL);
+  }
+
+  wrap?.once('close', () => this.destroy());
+
+  return handle;
+};
+
+function onSocketCloseDestroySSL(this: TLSSocket): void {
+  // We call destroySSL directly because it is already an async process.
+  this._destroySSL();
+  this[kRes] = null;
+}
+
+TLSSocket.prototype._destroySSL = function _destroySSL(this: TLSSocket): void {
+  // We disable floating promises rule because we don't want to change the
+  // function signature and return Promise<void>
+  //
+  // eslint-disable-next-line @typescript-eslint/no-floating-promises
+  this._handle?.socket.close().then(() => {
+    this[kPendingSession] = null;
+    this[kIsVerified] = false;
+  });
+};
+
+TLSSocket.prototype._init = function _init(this: TLSSocket): void {
+  const options = this._tlsOptions;
+
+  this.on('error', onerror.bind(this));
+
+  // This is thrown from net.Socket class.
```

Comment:
Nit "thrown" is typically used only for errors. "emitted" is likely the better word choice.